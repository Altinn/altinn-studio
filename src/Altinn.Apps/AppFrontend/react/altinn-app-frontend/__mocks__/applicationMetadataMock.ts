const applicationMetadataJSON = `{
  "id": "mockOrg/test-app",
  "org": "mockOrg",
  "title": {
    "nb": "Test App"
  },
  "dataTypes": [
    {
      "id": "test-data-model",
      "allowedContentTypes": [
        "application/xml"
      ],
      "appLogic": {
        "autoCreate": true,
        "classRef": "Altinn.App.Models.Skjema"
      },
      "taskId": "Task_1",
      "maxCount": 1,
      "minCount": 1
    },
    {
      "id": "ref-data-as-pdf",
      "allowedContentTypes": [
        "application/pdf"
      ],
      "maxCount": 0,
      "minCount": 0
    },
    {
      "id": "test-data-type-1",
      "allowedContentTypes": [
        "application/pdf"
      ],
      "maxCount": 5,
      "minCount": 0
    },
    {
      "id": "test-data-type-2",
      "allowedContentTypes": [
        "application/pdf"
      ],
      "maxCount": 2,
      "minCount": 0
    }
  ],
  "partyTypesAllowed": {
    "bankruptcyEstate": false,
    "organisation": false,
    "person": true,
    "subUnit": false
  },
  "created": "2020-06-29T08:47:12.425551Z",
  "createdBy": "test testesen",
  "lastChanged": "2020-06-29T08:47:12.4255537Z",
  "lastChangedBy": "test testesen"
}`;

export const applicationMetadataMock = JSON.parse(applicationMetadataJSON);
