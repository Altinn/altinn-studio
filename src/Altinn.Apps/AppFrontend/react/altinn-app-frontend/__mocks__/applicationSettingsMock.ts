const applicationSettingsJSON = `{
  "id": "mockOrg/test-app",
  "org": "mockOrg",
  "title": {
    "nb": "Test App"
  }}`;

export const applicationSettingsMock = JSON.parse(applicationSettingsJSON);
