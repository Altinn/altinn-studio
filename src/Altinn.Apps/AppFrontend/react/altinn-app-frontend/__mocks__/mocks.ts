export { getFormLayoutStateMock } from './formLayoutStateMock';
export { getInitialStateMock } from './initialStateMock';
export { getFormDataStateMock } from './formDataStateMock';
export { getLanguageFromCode as getLanguageMock } from '../../shared/src/language';
export { applicationMetadataMock } from './applicationMetadataMock';
