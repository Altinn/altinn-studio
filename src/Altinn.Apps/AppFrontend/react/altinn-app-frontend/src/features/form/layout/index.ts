import { IOption, Triggers } from '../../../types';

export interface ILayouts {
  [id: string]: ILayout;
}

export interface ILayoutEntry {
  id: string;
  type?: string;
}

export interface ILayoutGroup extends ILayoutEntry {
  children: string[];
  dataModelBindings?: IDataModelBindings;
  maxCount: number;
  textResourceBindings?: ITextResourceBindings;
  tableHeaders?: string[];
}

export interface ILayoutComponent extends ILayoutEntry {
  type: string;
  dataModelBindings: IDataModelBindings;
  isValid?: boolean;
  readOnly: boolean;
  disabled?: boolean;
  required: boolean;
  textResourceBindings: ITextResourceBindings;
  triggers?: Triggers[];
}
export interface IDataModelBindings {
  [id: string]: string;
}

export interface ITextResourceBindings {
  [id: string]: string;
}

export type ILayout = Array<ILayoutComponent | ILayoutGroup>;

export interface ISelectionComponentProps extends ILayoutComponent {
  options?: IOption[];
  optionsId?: string;
}

export interface IDatepickerProps extends ILayoutComponent { }

export interface IFileuploadProps extends ILayoutComponent {
  maxNumberOfAttachments: number;
  maxFileSizeInMB: number;
  displayMode: any;
  hasCustomFileEndings: boolean;
  validFileEndings: any;
}

declare enum HeaderSize {
  S,
  M,
  L,
}

export interface IHeaderProps extends ILayoutComponent {
  size: HeaderSize;
}

declare enum InputFieldType {
  text,
  email,
  password,
}

export interface IInputProps extends ILayoutComponent {
  inputType: InputFieldType;
}

export interface INavigationButtonProps extends ILayoutComponent {
  next?: string;
  previous?: string;
}

export interface IParagraphProps extends ILayoutComponent { }

export interface IRadioButtonsProps extends ILayoutComponent {
  options: IOption[];
  preselectedOptionIndex: number;
}

export interface ITextAreaProps extends ILayoutComponent { }

export interface IAdressComponent extends ILayoutComponent {
  addressTextResourceBinding: string;
  areaCodeTextResourceBinding: string;
  coTextResourceBinding: string;
  simpleDisplayMode: boolean;
}
