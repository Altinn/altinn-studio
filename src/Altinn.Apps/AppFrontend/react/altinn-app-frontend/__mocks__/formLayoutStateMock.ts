import { ILayoutState } from '../src/features/form/layout/formLayoutReducer';

export function getFormLayoutStateMock(customStates?: Partial<ILayoutState>): ILayoutState {
  const mockFormLayoutState: ILayoutState = {
    layouts: {
      FormLayout: [
        {
          id: 'field1',
          type: 'Input',
          dataModelBindings: {
            simple: 'Group.prop1',
          },
          textResourceBindings: {
            title: 'Title',
          },
          readOnly: false,
          required: false,
          disabled: false,
        },
        {
          id: 'field2',
          type: 'Input',
          dataModelBindings: {
            simple: 'Group.prop2',
          },
          textResourceBindings: {
            title: 'Title',
          },
          readOnly: false,
          required: false,
          disabled: false,
        },
        {
          id: 'field3',
          type: 'Input',
          dataModelBindings: {
            simple: 'Group.prop3',
          },
          textResourceBindings: {
            title: 'Title',
          },
          readOnly: false,
          required: false,
          disabled: false,
        },
      ],
    },
    error: {
      message: null,
      name: null,
    },
    uiConfig: {
      autoSave: true,
      focus: null,
      hiddenFields: [],
      repeatingGroups: null,
      currentView: 'FormLayout',
      navigationConfig: {},
      layoutOrder: [],
    },
  };

  return {
    ...mockFormLayoutState,
    ...customStates,
  };
}
