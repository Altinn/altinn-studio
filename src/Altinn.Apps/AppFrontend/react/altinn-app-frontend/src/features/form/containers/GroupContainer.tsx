/* eslint-disable no-undef */
/* eslint-disable react/no-array-index-key */
/* eslint-disable max-len */
import React from 'react';
import { Grid, makeStyles, createMuiTheme, TableContainer, Table, TableHead, TableRow, TableBody, TableCell, IconButton, useMediaQuery } from '@material-ui/core';
import { AltinnButton } from 'altinn-shared/components';
import altinnAppTheme from 'altinn-shared/theme/altinnAppTheme';
import { useSelector } from 'react-redux';
import { getLanguageFromKey, getTextResourceByKey } from 'altinn-shared/utils';
import { componentHasValidations, repeatingGroupHasValidations } from 'src/utils/validation';
import ErrorPaper from 'src/components/message/ErrorPaper';
import { ILayoutComponent, ILayoutGroup, ISelectionComponentProps } from '../layout';
import { renderGenericComponent } from '../../../utils/layout';
import FormLayoutActions from '../layout/formLayoutActions';
import { IRuntimeState, ITextResource, IRepeatingGroups, IValidations, IOption } from '../../../types';
import { IFormData } from '../data/formDataReducer';

export interface IGroupProps {
  id: string;
  container: ILayoutGroup;
  components: ILayoutComponent[]
}

const theme = createMuiTheme(altinnAppTheme);

const useStyles = makeStyles({
  addButton: {
    backgroundColor: theme.altinnPalette.primary.white,
    border: `2px dotted ${theme.altinnPalette.primary.blueMedium}`,
    color: theme.altinnPalette.primary.black,
    fontWeight: 'bold',
    width: '100%',
    '&:hover': {
      cursor: 'pointer',
      borderStyle: 'solid',
    },
    '&:focus': {
      outline: `2px solid ${theme.altinnPalette.primary.blueMedium}`,
      border: `none`,
      outlineOffset: 0,
      outlineColor: theme.altinnPalette.primary.blueMedium,
    },
  },
  addButtonText: {
    fontWeight: 400,
    fontSize: '1.6rem',
    borderBottom: `2px solid${theme.altinnPalette.primary.blue}`,
  },
  editContainer: {
    border: `2px dotted ${theme.altinnPalette.primary.blueMedium}`,
    padding: '12px',
  },
  table: {
    tableLayout: 'fixed',
    marginBottom: '12px',
  },
  tableHeader: {
    borderBottom: `2px solid ${theme.altinnPalette.primary.blueMedium}`,
    '& th': {
      fontSize: '1.4rem',
      padding: '0px',
      paddingLeft: '6px',
    },
  },
  tableBody: {
    '& td': {
      borderBottom: `2px dotted ${theme.altinnPalette.primary.blueMedium}`,
      padding: '0px',
      paddingLeft: '6px',
      fontSize: '1.4rem',
      whiteSpace: 'nowrap',
      textOverflow: 'ellipsis',
      overflow: 'hidden',
    },
  },
  tableRowError: {
    backgroundColor: '#F9CAD3;',
  },
  errorIcon: {
    fontSize: '2em',
    minWidth: '0px',
    minHeight: '0px',
    width: 'auto',
  },
  addIcon: {
    transform: 'rotate(45deg)',
    fontSize: '3.4rem',
    marginRight: '0.7rem',
  },
  deleteButton: {
    padding: '0px',
    color: 'black',
  },
  editIcon: {
    paddingLeft: '6px',
  },
  mobileGrid: {
    borderBottom: `2px dotted ${theme.altinnPalette.primary.blueMedium}`,
    paddingLeft: '0.6rem',
    paddingRight: '0.6rem',
  },
  mobileContainer: {
    borderTop: `2px solid ${theme.altinnPalette.primary.blueMedium}`,
    marginBottom: '1.2rem',
  },
  mobileText: {
    fontWeight: 500,
    float: 'left',
    maxWidth: '50%',
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',
    overflow: 'hidden',
  },
  mobileValueText: {
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',
    overflow: 'hidden',
    maxWidth: '50%',
    minWidth: '50%',
  },
  textContainer: {
    width: '100%',
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',
    overflow: 'hidden',
  },
});

export function GroupContainer({
  id,
  container,
  components,
}: IGroupProps): JSX.Element {
  const classes = useStyles();
  const renderComponents: ILayoutComponent[] = JSON.parse(JSON.stringify(components));
  const validations: IValidations = useSelector((state: IRuntimeState) => state.formValidations.validations);
  const language: any = useSelector((state: IRuntimeState) => state.language.language);
  const repeatingGroups: IRepeatingGroups = useSelector((state: IRuntimeState) => state.formLayout.uiConfig.repeatingGroups);
  const formData: IFormData = useSelector((state: IRuntimeState) => state.formData.formData);
  const [editIndex, setEditIndex] = React.useState<number>(-1);
  const options = useSelector((state: IRuntimeState) => state.optionState.options);
  const textResources: ITextResource[] = useSelector((state: IRuntimeState) => state.textResources.resources);
  const getRepeatingGroupIndex = (containerId: string) => {
    if (repeatingGroups && repeatingGroups[containerId]) {
      return repeatingGroups[containerId].count;
    }
    return -1;
  };
  const repeatinGroupIndex = getRepeatingGroupIndex(id);
  const tableHeaderComponents = container.tableHeaders || container.children || [];
  const mobileView = useMediaQuery('(max-width:992px)'); // breakpoint on altinn-modal
  const tableHasErrors = repeatingGroupHasValidations(validations, repeatinGroupIndex + 1, components);
  const componentTitles: string[] = [];
  renderComponents.forEach((component: ILayoutComponent) => {
    if (tableHeaderComponents.includes(component.id)) {
      componentTitles.push(component.textResourceBindings.title);
    }
  });

  const onClickAdd = () => {
    FormLayoutActions.updateRepeatingGroups(id);
    setEditIndex(repeatinGroupIndex + 1);
  };

  const onKeypressAdd = (event: React.KeyboardEvent<HTMLDivElement>) => {
    if (event.key === 'Enter') {
      onClickAdd();
    }
  };

  const getFormDataForComponent = (component: ILayoutComponent, index: number): string => {
    const dataModelBinding = (component.type === 'AddressComponent') ? component.dataModelBindings?.address : component.dataModelBindings?.simpleBinding;
    const replaced = dataModelBinding.replace(container.dataModelBindings.group, `${container.dataModelBindings.group}[${index}]`);
    if (component.type === 'Dropdown' || component.type === 'Checkboxes' || component.type === 'RadioButtons') {
      const selectionComponent = component as ISelectionComponentProps;
      let label: string;
      if (selectionComponent?.options) {
        label = selectionComponent.options.find((option: IOption) => option.value === formData[replaced])?.label;
      } else if (selectionComponent.optionsId) {
        label = options[selectionComponent.optionsId]?.find((option: IOption) => option.value === formData[replaced])?.label;
      }
      return getTextResourceByKey(label, textResources);
    }
    return formData[replaced] || '';
  };

  const onClickRemove = (groupIndex: number) => {
    setEditIndex(-1);
    FormLayoutActions.updateRepeatingGroups(id, true, groupIndex);
  };

  const onClickEdit = (groupIndex: number) => {
    if (groupIndex === editIndex) {
      setEditIndex(-1);
    } else {
      setEditIndex(groupIndex);
    }
  };

  const createRepeatingGroupComponents = () => {
    const componentArray = [];
    for (let i = 0; i <= repeatinGroupIndex; i++) {
      const childComponents = renderComponents.map((component: ILayoutComponent) => {
        const componentDeepCopy: ILayoutComponent = JSON.parse(JSON.stringify(component));
        const dataModelBindings = { ...componentDeepCopy.dataModelBindings };

        const groupDataModelBinding = container.dataModelBindings.group;
        Object.keys(dataModelBindings).forEach((key) => {
          // eslint-disable-next-line no-param-reassign
          dataModelBindings[key] = dataModelBindings[key].replace(groupDataModelBinding, `${groupDataModelBinding}[${i}]`);
        });
        const deepCopyId = `${componentDeepCopy.id}-${i}`;
        return {
          ...componentDeepCopy,
          dataModelBindings,
          id: deepCopyId,
          baseComponentId: componentDeepCopy.id,
        };
      });
      componentArray.push(childComponents);
    }
    return componentArray;
  };

  const repeatingGroupDeepCopyComponents = createRepeatingGroupComponents();

  return (
    <>
      <Grid
        container={true}
        data-testid={`group-${id}`}
        id={`group-${id}`}
      >
        {!mobileView &&
        <TableContainer component={Grid}>
          <Table className={classes.table}>
            <TableHead className={classes.tableHeader}>
              <TableRow>
                {componentTitles.map((title: string) => (
                  <TableCell align='left' key={title}>
                    {getTextResourceByKey(title, textResources)}
                  </TableCell>
                ))}
                <TableCell/>
              </TableRow>
            </TableHead>
            <TableBody className={classes.tableBody}>
              {(repeatinGroupIndex >= 0) && [...Array(repeatinGroupIndex + 1)].map((_x: any, repeatingGroupIndex: number) => {
                const rowHasErrors = components.some((component: ILayoutComponent) => {
                  return componentHasValidations(validations, `${component.id}-${repeatingGroupIndex}`);
                });
                return (
                  <TableRow className={rowHasErrors ? classes.tableRowError : ''} key={repeatingGroupIndex}>
                    {components.map((component: ILayoutComponent) => {
                      if (!tableHeaderComponents.includes(component.id)) {
                        return null;
                      }
                      return (
                        <TableCell key={`${component.id} ${repeatingGroupIndex}`}>
                          {getFormDataForComponent(component, repeatingGroupIndex)}
                        </TableCell>
                      );
                    })}
                    <TableCell align='right' key={`delete-${repeatingGroupIndex}`}>
                      <IconButton style={{ color: 'black' }} onClick={() => onClickEdit(repeatingGroupIndex)}>
                        {rowHasErrors ?
                          getLanguageFromKey('general.edit_alt_error', language) :
                          getLanguageFromKey('general.edit_alt', language)}
                        <i className={rowHasErrors ?
                          `ai ai-circle-exclamation a-icon ${classes.errorIcon} ${classes.editIcon}` :
                          `fa fa-editing-file ${classes.editIcon}`}
                        />
                      </IconButton>
                    </TableCell>
                  </TableRow>);
              })}
            </TableBody>
          </Table>
        </TableContainer>}
        {mobileView &&
        <Grid
          container={true} direction='column'
          className={classes.mobileContainer}
        >
          {(repeatinGroupIndex >= 0) && [...Array(repeatinGroupIndex + 1)].map((_x: any, repeatingGroupIndex: number) => {
            const rowHasErrors = components.some((component: ILayoutComponent) => {
              return componentHasValidations(validations, `${component.id}-${repeatingGroupIndex}`);
            });
            return (
              <Grid
                item={true} container={true}
                justify='flex-end' direction='row'
                className={`${classes.mobileGrid} ${rowHasErrors ? classes.tableRowError : ''}`}
              >
                <Grid item={true}>
                  <IconButton
                    style={{
                      color: 'black', padding: '0px', paddingLeft: '6px',
                    }} onClick={() => onClickEdit(repeatingGroupIndex)}
                  >
                    {rowHasErrors ?
                      getLanguageFromKey('general.edit_alt_error', language) :
                      getLanguageFromKey('general.edit_alt', language)}
                    <i className={rowHasErrors ?
                      `ai ai-circle-exclamation ${classes.errorIcon}` :
                      `fa fa-editing-file ${classes.editIcon}`}
                    />
                  </IconButton>
                </Grid>
                {componentTitles.map((title: string, index: number) => {
                  return (
                    <Grid item={true} className={rowHasErrors ? `${classes.tableRowError} ${classes.textContainer}` : classes.textContainer}>
                      <div className={classes.mobileText}>
                        {`${getTextResourceByKey(title, textResources)}`}
                      </div>
                      <div
                        className={classes.mobileValueText}
                      >
                        {`: ${getFormDataForComponent(components[index], repeatingGroupIndex)}`}
                      </div>
                    </Grid>
                  );
                })}
              </Grid>
            );
          })}
        </Grid>
        }
      </Grid>
      <Grid
        container={true}
        justify='flex-end'
      />
      {((editIndex < 0) && ((repeatinGroupIndex + 1) < container.maxCount)) &&
      <Grid
        container={true}
        direction='row'
        justify='center'
      >
        <Grid
          item={true}
          container={true}
          direction='row'
          xs={12}
          className={classes.addButton}
          role='button'
          tabIndex={0}
          onClick={onClickAdd}
          onKeyPress={(event) => onKeypressAdd(event)}
          justify='center'
          alignItems='center'
        >
          <Grid item={true}>
            <i className={`fa fa-exit ${classes.addIcon}`} />
          </Grid>
          <Grid item={true}>
            <span className={classes.addButtonText}>
              {`${getLanguageFromKey('general.add_new', language)} 
              ${container.textResourceBindings?.add_button ? getTextResourceByKey(container.textResourceBindings.add_button, textResources) : ''}`}
            </span>
          </Grid>
        </Grid>
      </Grid>
      }
      {(editIndex >= 0) &&
      <Grid container={true} className={classes.editContainer}>
        <Grid
          container={true}
          direction='column'
        >
          <Grid
            item={true}
            container={true}
            direction='column'
            alignItems='flex-end'
          >
            <Grid item={true}>
              <IconButton
                classes={{ root: classes.deleteButton }}
                onClick={() => onClickRemove(editIndex)}
              >
                {getLanguageFromKey('general.delete', language)}
                <i className='ai ai-trash'/>
              </IconButton>
            </Grid>
          </Grid>
          <Grid item={true} xs={12}>
            { repeatingGroupDeepCopyComponents[editIndex].map(renderGenericComponent) }
          </Grid>
          <Grid item={true} xs={12}>
            <AltinnButton
              btnText={getLanguageFromKey('general.save', language)}
              onClickFunction={() => setEditIndex(-1)}
              id={`add-button-grp-${id}`}
            />
          </Grid>
        </Grid>
      </Grid>}
      {tableHasErrors &&
      <Grid container={true} style={{ paddingTop: '12px' }}>
        <ErrorPaper
          message={getLanguageFromKey('group.row_error', language)}
        />
      </Grid>
      }
    </>
  );
}
