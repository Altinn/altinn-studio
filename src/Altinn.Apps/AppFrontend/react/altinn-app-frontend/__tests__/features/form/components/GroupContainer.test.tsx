/* eslint-disable max-len */
/* eslint-disable no-undef */
/* tslint:disable:jsx-wrap-multiline */
import 'jest';
import * as React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import * as renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { GroupContainer } from '../../../../src/features/form/containers/GroupContainer';
import { getInitialStateMock } from '../../../../__mocks__/mocks';
import { ILayoutGroup } from '../../../../src/features/form/layout';

describe('>>> features/form/components/Group.tsx', () => {
  let mockStore: any;
  let mockLayout: any;
  let mockData: any;
  let mockComponents: any;
  let mockTextResources: any;
  let mockContainer: ILayoutGroup;

  beforeAll(() => {
    window.matchMedia = jest.fn().mockImplementation((query) => {
      return {
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
      };
    });
  });

  beforeEach(() => {
    const createStore = configureStore();
    mockComponents = [
      {
        id: 'field1',
        type: 'Input',
        dataModelBindings: {
          simpleBinding: 'Group.prop1',
        },
        textResourceBindings: {
          title: 'Title1',
        },
        readOnly: false,
        required: false,
        disabled: false,
      },
      {
        id: 'field2',
        type: 'Input',
        dataModelBindings: {
          simpleBinding: 'Group.prop2',
        },
        textResourceBindings: {
          title: 'Title2',
        },
        readOnly: false,
        required: false,
        disabled: false,
      },
      {
        id: 'field3',
        type: 'Input',
        dataModelBindings: {
          simpleBinding: 'Group.prop3',
        },
        textResourceBindings: {
          title: 'Title3',
        },
        readOnly: false,
        required: false,
        disabled: false,
      },
      {
        id: 'field4',
        type: 'Checkboxes',
        dataModelBindings: {
          simpleBinding: 'some-group.checkboxBinding',
        },
        textResourceBindings: {
          title: 'Title4',
        },
        readOnly: false,
        required: false,
        disabled: false,
        options: [{ value: 'option.value', label: 'option.label' }],
      },
    ];

    mockLayout = {
      layout: [
        {
          id: 'mock-container-id',
          type: 'group',
          dataModelBindings: {
            group: 'Group',
          },
          children: [
            'field1',
            'field2',
            'field3',
            'field4',
          ],
        },
      ].concat(mockComponents),
      uiConfig: {
        hiddenFields: [],
        repeatingGroups: {
          'mock-container-id': {
            count: 3,
          },
        },
        autosave: false,
      },
    };

    mockContainer = {
      id: 'mock-container-id',
      children: [
        'field1',
        'field2',
        'field3',
        'field4',
      ],
      maxCount: 8,
      dataModelBindings: {
        group: 'some-group',
      },
    };

    mockData = {
      formData: {
        'some-group[1].checkboxBinding': 'option.value',
      },
    };

    mockTextResources = {
      resources: [{ id: 'option.label', value: 'Value to be shown' }],
    };

    const initialState = getInitialStateMock({ formLayout: mockLayout, formData: mockData, textResources: mockTextResources});
    mockStore = createStore(initialState);
  });

  it('+++ should match snapshot', () => {
    const rendered = renderer.create(
      <Provider store={mockStore}>
        <GroupContainer
          components={mockComponents}
          id='testGroupId'
          key='testKey'
          container={mockContainer}
        />
      </Provider>,
    );
    expect(rendered).toMatchSnapshot();
  });

  it('+++ should render add new button', async () => {
    const utils = render(
      <Provider store={mockStore}>
        <GroupContainer
          components={mockComponents}
          container={mockContainer}
          id='testGroupId'
          key='testKey'
        />
      </Provider>,
    );
    const item = await utils.findByText('Legg til ny');
    expect(item).not.toBe(null);
  });

  it('+++ should render add new button with custom label when supplied', async () => {
    const mockContainerWithLabel: ILayoutGroup = {
      textResourceBindings: {
        add_button: 'person',
      },
      ...mockContainer,
    };
    const utils = render(
      <Provider store={mockStore}>
        <GroupContainer
          container={mockContainerWithLabel}
          components={mockComponents}
          id='testGroupId'
          key='testKey'
        />
      </Provider>,
    );
    const item = await utils.findByText('Legg til ny person');
    expect(item).not.toBeNull();
  });

  it('+++ should not show add button when maxOccurs is reached', () => {
    const mockContainerWithMaxCount = {
      ...mockContainer,
      maxCount: 3,
    };
    const utils = render(
      <Provider store={mockStore}>
        <GroupContainer
          components={mockComponents}
          container={mockContainerWithMaxCount}
          id='mock-container-id'
          key='testKey'
        />
      </Provider>,
    );
    const addButton = utils.queryByText('Legg til ny');
    expect(addButton).toBeNull();
  });

  it('+++ should show option label when displaying selection components', async () => {
    const utils = render(
      <Provider store={mockStore}>
        <GroupContainer
          components={mockComponents}
          container={mockContainer}
          id='mock-container-id'
          key='testKey'
        />
      </Provider>,
    );
    const item = await utils.findByText('Value to be shown');
    expect(item).not.toBeNull();
  });
});
