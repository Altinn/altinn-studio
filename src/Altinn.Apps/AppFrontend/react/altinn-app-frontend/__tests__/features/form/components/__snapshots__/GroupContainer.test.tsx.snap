// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> features/form/components/Group.tsx +++ should match snapshot 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-item"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-item"
    data-testid="group-mock-container-id"
    id="group-mock-container-id"
  >
    <div
      className="MuiGrid-root MuiTableContainer-root"
    >
      <table
        className="MuiTable-root makeStyles-table-1"
        role={null}
      >
        <thead
          className="MuiTableHead-root makeStyles-tableHeader-2"
          role={null}
        >
          <tr
            className="MuiTableRow-root MuiTableRow-head"
            role={null}
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft"
              scope="col"
            >
              Title1
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft"
              scope="col"
            >
              Title2
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft"
              scope="col"
            >
              Title3
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft"
              scope="col"
            >
              Title4
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head"
              scope="col"
            />
          </tr>
        </thead>
        <tbody
          className="MuiTableBody-root makeStyles-tableBody-3"
          role={null}
        >
          <tr
            className="MuiTableRow-root"
            role={null}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
            >
              <button
                className="MuiButtonBase-root MuiIconButton-root"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                style={
                  Object {
                    "color": "black",
                  }
                }
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  Rediger
                  <i
                    className="fa fa-editing-file makeStyles-editIcon-6"
                  />
                </span>
                <span
                  className="MuiTouchRipple-root"
                />
              </button>
            </td>
          </tr>
          <tr
            className="MuiTableRow-root"
            role={null}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              Value to be shown
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
            >
              <button
                className="MuiButtonBase-root MuiIconButton-root"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                style={
                  Object {
                    "color": "black",
                  }
                }
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  Rediger
                  <i
                    className="fa fa-editing-file makeStyles-editIcon-6"
                  />
                </span>
                <span
                  className="MuiTouchRipple-root"
                />
              </button>
            </td>
          </tr>
          <tr
            className="MuiTableRow-root"
            role={null}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
            >
              <button
                className="MuiButtonBase-root MuiIconButton-root"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                style={
                  Object {
                    "color": "black",
                  }
                }
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  Rediger
                  <i
                    className="fa fa-editing-file makeStyles-editIcon-6"
                  />
                </span>
                <span
                  className="MuiTouchRipple-root"
                />
              </button>
            </td>
          </tr>
          <tr
            className="MuiTableRow-root"
            role={null}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
            >
              
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
            >
              <button
                className="MuiButtonBase-root MuiIconButton-root"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                style={
                  Object {
                    "color": "black",
                  }
                }
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  Rediger
                  <i
                    className="fa fa-editing-file makeStyles-editIcon-6"
                  />
                </span>
                <span
                  className="MuiTouchRipple-root"
                />
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-flex-end"
  />
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-center"
  >
    <div
      className="MuiGrid-root makeStyles-addButton-12 MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-center MuiGrid-grid-xs-12"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiGrid-root MuiGrid-item"
      >
        <i
          className="fa fa-exit makeStyles-addIcon-14"
        />
      </div>
      <div
        className="MuiGrid-root MuiGrid-item"
      >
        <span
          className="makeStyles-addButtonText-13"
        >
          Legg til ny
            
        </span>
      </div>
    </div>
  </div>
</div>
`;
